ms2pip/ms2pip_tools/spectrum_output.py

```diff
@@ -663,6 +666,77 @@ def write_bibliospec(self):
+        Write MS2PIP predictions to a DLIB SQLite file.
+        """
+        from ms2pip.ms2pip_tools.dlib import Entry, PeptideToProtein, open_sqlite, metadata
+
+        normalization = "basepeak_10000"
+        precision = 5
+        if not self.normalization == normalization:
+            self._normalize_spectra(method=normalization)
+            self._generate_preds_dict()
+        if not self.peprec_dict:
+            self._generate_peprec_dict()
+        if precision not in self.diff_modification_mapping:
+            self._generate_diff_modification_mapping(precision)
+
+        filename = "{}.dlib".format(self.output_filename)
+        logger.info("writing results to %s", filename)
+
+        logger.warn("write mode is ignored for DLIB at the file mode, although append or not is respected")
+        if 'a' not in self.write_mode and os.path.exists(filename):
+            os.remove(filename)
+
+        if self.return_stringbuffer:
+            raise NotImplementedError("`return_stringbuffer` not implemented for DLIB output.")
+
+        if not self.has_rt:
+            raise NotImplementedError("Retention times required to write DLIB file.")
+
+        with open_sqlite(filename) as connection:
+            metadata.create_all()
+
+            for spec_id, peprec in self.peprec_dict.items():
+                seq = peprec["peptide"]
+                mods = peprec["modifications"]
+                charge = peprec["charge"]
+
+                prec_mass, prec_mz = self.mods.calc_precursor_mz(seq, mods, charge)
+                mod_seq = self._get_diff_modified_sequence(seq, mods, precision=precision)
+
+                all_peaks = sorted(itertools.chain.from_iterable(self.preds_dict[spec_id]['peaks'].values()), key=itemgetter(1))
+                mzs = [peak[1] for peak in all_peaks]
+                intensities = [peak[2] for peak in all_peaks]
+
+                connection.execute(Entry.insert().values(
+                    PrecursorMz=prec_mz,
+                    PrecursorCharge=charge,
+                    PeptideModSeq=mod_seq,
+                    PeptideSeq=seq,
+                    Copies=1,
+                    RTInSeconds=peprec["rt"],
+                    Score=0,
+                    MassEncodedLength=len(mzs),
+                    MassArray=mzs,
+                    IntensityEncodedLength=len(intensities),
+                    IntensityArray=intensities,
+                    SourceFile=self.output_filename  # TODO: any alternative?
+                ))
+
+                if self.has_protein_list:
+                    protein_list = peprec["protein_list"]
+                    if isinstance(protein_list, str):
+                        protein_list = literal_eval(protein_list)
+
+                    for protein in protein_list:
+                        connection.execute(PeptideToProtein.insert().values(
+                            PeptideSeq=seq,
+                            ProteinAccession=protein
+                        ))
+
```

Comment:
Is the PeptideToProtein table required in the DLIB file? If so, maybe we should raise an error, or at least log a warning that the DLIB will not be compliant?

Reply:
The table is, but there is [no mention](https://bitbucket.org/searleb/encyclopedia/wiki/EncyclopeDIA%20File%20Formats) there has to be data.